lib/src/core/comms/secure_channel.rs

```diff
@@ -1048,8 +1066,33 @@ impl SecureChannel {
         self.local_keys.as_ref().unwrap()
     }
 
-    fn remote_keys(&self) -> &(Vec<u8>, AesKey, Vec<u8>) {
-        self.remote_keys.as_ref().unwrap()
+    fn insert_remote_keys(&mut self, keys: (Vec<u8>, AesKey, Vec<u8>)) {
+        // First remove any expired keys.
+        self.remote_keys
+            .retain(|_, v| DateTime::now() < v.expires_at);
+
+        let expires_at = (self.token_lifetime as f32 * 1.25).ceil();
+        let expires_at = Duration::milliseconds(expires_at as i64);
+
+        // Then insert the new keys to ensure there is
+        // always at least one set of keys available.
+        self.remote_keys.insert(
+            self.token_id,
+            RemoteKeys {
+                keys,
+                expires_at: self.token_created_at + expires_at,
```

Comment:
Without looking too closely at the code, it seems like `token_created_at` is subject to some internal invariants that could be better modeled (it must correspond to the expires_at in the remote_keys entry given by `token_id`). Perhaps a target for some future refactoring...

Not a blocker.

Reply:
I'm not sure I understand what you mean with `token_created_at is subject to some internal invariants`? Both the token ID and associated created at time are set using the same call during the issue or renew of a channel.

Reply:
As I said, not a big deal, my point is just that there's now a built-in relationship in the `SecureChannel` type between `token_id`, `token_created_at`, and `RemoteKeys.expires_at`, it would be better to not have the `token_created_at` parameter at all, and instead store it only inside the `remote_keys` map.

Reply:
As I said, the whole `SecureChannel` type has a few of these invariants that could be modeled in the type system, IMO. It can be done in a future PR, I'm fine with this as is now.